File: core/src/com/desitum/library/drawing/Drawable.java
package com.desitum.library.drawing;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.NinePatch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.scenes.scene2d.utils.NinePatchDrawable;

/**
 * Created by kodyvanry on 5/1/17.
 */

public class Drawable {

    private TextureRegion mTextureRegion;
    private NinePatch mNinePatch;
    private Color mColor;

    public Drawable(Texture texture) {
        this.mTextureRegion = new TextureRegion(texture);
    }

    public Drawable(TextureRegion textureRegion) {
        this.mTextureRegion = textureRegion;
    }

    public Drawable(NinePatch ninePatch) {
        this.mNinePatch = ninePatch;
    }

    public static Drawable loadDrawable(String filePath) {
        return loadDrawable(new Texture(filePath), false);
    }

    public static Drawable loadDrawable(String filePath, boolean isNinePatch) {
        return loadDrawable(new Texture(filePath), isNinePatch);
    }

    public static Drawable loadDrawable(Texture texture, boolean isNinePatch) {
        if (isNinePatch) {
            int width = texture.getWidth();
            int height = texture.getHeight();
            return new Drawable(
                    new NinePatch(
                            texture,
                            width / 3,
                            width / 3,
                            height / 3,
                            height / 3
                    )
            );
        }
        return new Drawable(texture);
    }

    public void setColor(Color color) {
        this.mColor = color;
        if (mNinePatch != null) {
            mNinePatch = new NinePatch(mNinePatch, color);
        }
    }

    public void draw(Batch batch, float x, float y, float width, float height) {
        if (mNinePatch != null) {
            mNinePatch.draw(batch, x, y, Math.max(width, mNinePatch.getTotalWidth()), Math.max(height, mNinePatch.getTotalHeight()));
        } else if (mColor != null && mTextureRegion != null) {
            Color originalColor = batch.getColor();
            batch.setColor(mColor);
            batch.draw(mTextureRegion, x, y, width, height);
            batch.setColor(originalColor);
        } else {
            batch.draw(mTextureRegion, x, y, width, height);
        }
    }
}

File: core/src/com/desitum/library/drawing/Drawable.kt
package com.desitum.library.drawing

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.NinePatch
import com.badlogic.gdx.graphics.g2d.TextureRegion

/**
 * Created by kodyvanry on 5/1/17.
 */

class Drawable {

    // THESE SHOULD BE COPIED WHEN GETTING A DRAWABLE FROM A DRAWABLE
    var textureRegion: TextureRegion? = null
        private set
    var ninePatch: NinePatch? = null
        private set
    var color: Color = Color.WHITE
        set(value) {
            ninePatch?.let { ninePatch = NinePatch(it, value) }
            field = value
        }

    val width: Float
        get() = textureRegion?.regionWidth?.plus(0f) ?: ninePatch?.totalWidth!!

    val height: Float
        get() = textureRegion?.regionWidth?.plus(0f) ?: ninePatch?.totalWidth!!


    // THESE SHOULD **NOT** BE COPIED WHEN GETTING A DRAWABLE FROM A DRAWABLE
    var rotation = 0f

    constructor(texture: Texture) {
        this.textureRegion = TextureRegion(texture)
    }

    constructor(textureRegion: TextureRegion) {
        this.textureRegion = textureRegion
    }

    constructor(ninePatch: NinePatch) {
        this.ninePatch = ninePatch
    }

    constructor(drawable: Drawable) {
        this.textureRegion = drawable.textureRegion
        this.ninePatch = drawable.ninePatch
        this.color = drawable.color
    }

    @JvmOverloads
    fun draw(batch: Batch, x: Float, y: Float, width: Float, height: Float,
             originX: Float = width / 2, originY: Float = height / 2,
             scaleX: Float = 1f, scaleY: Float = 1f,
             rotation: Float = 0f) {
        ninePatch?.draw(batch, x, y, Math.max(width, ninePatch!!.totalWidth), Math.max(height, ninePatch!!.totalHeight))
        val originalColor = batch.color
        batch.color = color
        textureRegion?.let { batch.draw(it, x, y, originX, originY, width, height, scaleX, scaleY, rotation) }
        batch.color = originalColor
    }

    companion object {

        @JvmStatic fun loadDrawable(filePath: String): Drawable {
            return loadDrawable(Texture(filePath), false)
        }

        @JvmStatic fun loadDrawable(filePath: String, isNinePatch: Boolean): Drawable {
            return loadDrawable(Texture(filePath), isNinePatch)
        }

        @JvmStatic fun loadDrawable(texture: Texture, isNinePatch: Boolean): Drawable {
            if (isNinePatch) {
                val width = texture.width
                val height = texture.height
                return Drawable(
                        NinePatch(
                                texture,
                                width / 3,
                                width / 3,
                                height / 3,
                                height / 3
                        )
                )
            }
            return Drawable(texture)
        }
    }
}
